### ISSUE TITLE
[BUG][3002rc1] lvm.lv_present throws an exception when LV exists

### ISSUE THREAD

**Issue by dkacar-oradian** [saltstack/salt#58759]
**Description**
When used with `extents: 100%VG` `lvm.lv_present` creates logical volume without problems. But when I run the same state again I get an exception. Tested on CentOS 7, I'm not sure if that happens on other Linux distributions.

**Setup**
Create a virtual machine with one extra unused disk. In the following example that disk is `/dev/sdb`.

State which causes an exception is:
```
lv_vgbkp_bar00:
  lvm.lv_present:
  - name: bar00
  - vgname: vgbkp
  - extents: 100%VG
```

<details><summary>The whole test file</summary>

```
mount_data_dir:
  file.directory:
    - user: root
    - group: root
    - mode: 755
    - names:
      - /data

pv_dsk_/dev/sdb:
  lvm.pv_present:
    - name: /dev/sdb

vg_vgbkp:
  lvm.vg_present:
  - name: vgbkp
  - devices:
    - /dev/sdb
  - require:
    - pv_dsk_/dev/sdb

lv_vgbkp_bar00:
  lvm.lv_present:
  - name: bar00
  - vgname: vgbkp
  - extents: 100%VG

lv_vgbkp_bar00_format:
  blockdev.formatted:
  - name: /dev/mapper/vgbkp-bar00
  - fs_type: ext4
  - require:
    - lv_vgbkp_bar00

mount_barman:
  mount.mounted:
  - name: /data/barman
  - mkmnt: true
  - device: /dev/mapper/vgbkp-bar00
  - fstype: ext4
  - require:
    - mount_data_dir
```
</details>

**Steps to Reproduce the behavior**
Run the file.
```
salt stest-barman-01.xdc state.apply test
```
Everything is fine and the file system is created on the minion.
<details><summary>Salt output</summary>

```
stest-barman-01.xdc:
----------
          ID: mount_data_dir
    Function: file.directory
        Name: /data
      Result: True
     Comment: Directory /data updated
     Started: 14:18:50.664100
    Duration: 26.845 ms
     Changes:   
              ----------
              /data:
                  New Dir
----------
          ID: pv_dsk_/dev/sdb
    Function: lvm.pv_present
        Name: /dev/sdb
      Result: True
     Comment: Created Physical Volume /dev/sdb
     Started: 14:18:50.693774
    Duration: 280.726 ms
     Changes:   
              ----------
              created:
                  True
----------
          ID: vg_vgbkp
    Function: lvm.vg_present
        Name: vgbkp
      Result: True
     Comment: Created Volume Group vgbkp
     Started: 14:18:50.976765
    Duration: 141.945 ms
     Changes:   
              ----------
              created:
                  ----------
                  Output from vgcreate:
                      Volume group "vgbkp" successfully created
                  vgbkp:
                      ----------
                      Actual Physical Volumes:
                          1
                      Allocated Physical Extents:
                          0
                      Current Logical Volumes:
                          0
                      Current Physical Volumes:
                          1
                      Free Physical Extents:
                          7679
                      Internal Volume Group Number:
                          -1
                      Maximum Logical Volume Size:
                          -1
                      Maximum Logical Volumes:
                          0
                      Maximum Physical Volumes:
                          0
                      Open Logical Volumes:
                          0
                      Physical Extent Size (kB):
                          4096
                      Total Physical Extents:
                          7679
                      UUID:
                          lwqiA3-MWc5-NTxu-zZfo-Xn9Z-BtQq-uTDYeK
                      Volume Group Access:
                          r/w
                      Volume Group Name:
                          vgbkp
                      Volume Group Size (kB):
                          31453184
                      Volume Group Status:
                          772
----------
          ID: lv_vgbkp_bar00
    Function: lvm.lv_present
        Name: bar00
      Result: True
     Comment: Created Logical Volume bar00
     Started: 14:18:51.119482
    Duration: 154.302 ms
     Changes:   
              ----------
              created:
                  ----------
                  /dev/vgbkp/bar00:
                      ----------
                      Allocated Logical Extents:
                          -1
                      Allocation Policy:
                          0
                      Current Logical Extents Associated:
                          7679
                      Internal Logical Volume Number:
                          -1
                      Logical Volume Access:
                          3
                      Logical Volume Name:
                          /dev/vgbkp/bar00
                      Logical Volume Size:
                          62906368
                      Logical Volume Status:
                          1
                      Major Device Number:
                          253
                      Minor Device Number:
                          3
                      Open Logical Volumes:
                          0
                      Read Ahead Sectors:
                          -1
                      Volume Group Name:
                          vgbkp
                  Output from lvcreate:
                      Logical volume "bar00" created.
----------
          ID: lv_vgbkp_bar00_format
    Function: blockdev.formatted
        Name: /dev/mapper/vgbkp-bar00
      Result: True
     Comment: /dev/mapper/vgbkp-bar00 has been formatted with ext4
     Started: 14:18:51.279144
    Duration: 605.649 ms
     Changes:   
              ----------
              new:
                  ext4
              old:
                  ext4
----------
          ID: mount_barman
    Function: mount.mounted
        Name: /data/barman
      Result: True
     Comment: Target was successfully mounted. Added new entry to the fstab.
     Started: 14:18:51.887724
    Duration: 323.301 ms
     Changes:   
              ----------
              mount:
                  True
              persist:
                  new

Summary for stest-barman-01.xdc
------------
Succeeded: 6 (changed=6)
Failed:    0
------------
Total states run:     6
Total run time:   1.533 s
```
</details>

Now run it again:
```
# salt --state-verbose False stest-barman-01.xdc state.apply test                                           
stest-barman-01.xdc:
----------
          ID: lv_vgbkp_bar00
    Function: lvm.lv_present
        Name: bar00
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python3.6/site-packages/salt/state.py", line 2154, in call
                  *cdata["args"], **cdata["kwargs"]
                File "/usr/lib/python3.6/site-packages/salt/loader.py", line 2188, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python3.6/site-packages/salt/states/lvm.py", line 334, in lv_present
                  if force is False and (size_mb < old_size_mb or extents < old_extents):
              TypeError: '<' not supported between instances of 'str' and 'int'
     Started: 14:19:42.874347
    Duration: 43.475 ms
     Changes:   
----------
          ID: lv_vgbkp_bar00_format
    Function: blockdev.formatted
        Name: /dev/mapper/vgbkp-bar00
      Result: False
     Comment: One or more requisite failed: test.lv_vgbkp_bar00
     Started: 14:19:42.925035
    Duration: 0.008 ms
     Changes:   

Summary for stest-barman-01.xdc
------------
Succeeded: 4
Failed:    2
------------
Total states run:     6
Total run time: 274.294 ms
```

**Expected behavior**
The exception should not have happened. This wasn't happening in the previous releases.

**Screenshots**

**Versions Report**
<details><summary>salt --versions-report</summary>
Master is at the same version as the minion.

```
# salt-minion --versions-report
Salt Version:
           Salt: 3002rc1
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.11.1
        libgit2: Not Installed
       M2Crypto: 0.35.2
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: Not Installed
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.6.8 (default, Apr  2 2020, 13:34:55)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 17.0.0
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7 Core
         locale: ISO-8859-2
        machine: x86_64
        release: 3.10.0-1127.18.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7 Core
 
```
</details>

**Additional context**

**Comment by piterpunk** [comment]
Hi @dkacar-oradian ,

The [documentation](https://docs.saltstack.com/en/master/ref/states/all/salt.states.lvm.html#salt.states.lvm.lv_present) says that "extents" should be a number:
```
extents
    The number of logical extents to allocate
```
The "100%VG" works at first time because it is passed to lv_create without any treatment. The second time, it have a comparision between the existent amount of states (which is a number) and the "100%VG", which is a string.

Strictly speaking, the bug is in accepting the "100%VG" at first time.

**Comment by dkacar-oradian** [comment]
That severely breaks existing functionality which worked fine in a number of previous releases.

And makes Salt's lvm module not particularly useful.

With the command line utility `lvcreate` you can use `%VG`, `%FREE` and more esoteric ones like `%PVS` or `%ORIGIN`. And these things are very useful.

What am I supposed to do without those options? Calculate percentages myself or just ignore Salt's module and implement the whole thing with a stateful script?

**Comment by piterpunk** [comment]
I am only pointing that the state is working and was tested as specified in documentation. I did not wrote the documentation, but I agree with what is defined there. Allowing the percentages in the states brings some pratical problems: 

**How can be kept a state that enforces the "%FREE"?**

The user specify "50%FREE", the state ran and there is 30 extents free in the VG, the LV is created with 15 extents. What should the state does at second time? Resize the Logical Volume to 7 extents? 

**How can be kept a state that enforces the "%VG"?**

The percentages are not deterministic in LVM, you can specify "100%VG" and, if there is already another LV using some space, ended with a smaller LV. This is specified in LVM documentation:

> When expressed as a percentage, the size defines an upper limit for the number of logical extents in the new LV. The precise number of logical extents in the new LV is not determined until the command has completed.

And that is how it works:
```
root@bat:~# lvs testvg01
  LV        VG       Attr       LSize  Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  salttest  testvg01 -wi-ao---- 10.00g                                                    
  salttest2 testvg01 -wi-a-----  2.00g                                                    
root@bat:~# lvcreate --extents '100%VG' -n salttest3 /dev/testvg01
  Reducing 100%VG to remaining free space <4.00 GiB in VG.
  Logical volume "salttest3" created.
root@bat:~# lvs testvg01
  LV        VG       Attr       LSize  Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  salttest  testvg01 -wi-ao---- 10.00g                                                    
  salttest2 testvg01 -wi-a-----  2.00g                                                    
  salttest3 testvg01 -wi-a----- <4.00g                             
```
How the state should handle that salttest3, specified as 100%VG, have less than 25% of it?

I guess the only way to accept the percentages is, in reality, ignoring them. Pass the option to linux_lvm.lvcreate if the logical volume is not there and ignore it if the logical volume exists. This will bring back the previous behaviour, but the user should be warn in documentation that if he changes the size from '10%VG' to '30%VG' the logical volume will not be resized accordingly.

> And makes Salt's lvm module not particularly useful.

IMHO, the resize capability is much more useful than specifying the percentage. But probably we have different use cases.

**Comment by dkacar-oradian** [comment]
Exactly, do not resize if the size is a percentage.

And there should be a flag (per state and in minion config, at least) to disable resizing even if it's a number without weird suffixes. That is a very dangerous operation which shouldn't be attempted lightly, so administrators should have that option.

**Comment by piterpunk** [comment]
> Exactly, do not resize if the size is a percentage.

That can be easily done, and will preserve the old undocumented behaviour. I will only wait the @cmcmarrow input to begin coding.

> And there should be a flag (per state and in minion config, at least) to disable resizing even if it's a number without weird suffixes. That is a very dangerous operation which shouldn't be attempted lightly, so administrators should have that option.

The state already has the "force" flag, which defaults to False. This avoids data loss by an acidental logical volume shrink. There is also a "resizefs" flag, which enables / disables resizing the file system, but not the logical volume.

Keep a logical volume in the specified size is the expected behaviour of state that includes the size. If the user increases the amount of extents or the size of a LV in the state definition, the LV should be present with the defined amount extents or size.

I disagree that the logical volume resize is a dangerous operation, growing a filesystem is by far the most common LVM related request in the Unix/Linux team's queue. A logical volume is created only once, but extended until the application system that uses it is decommissioned, which could be close to forever.

**Comment by dkacar-oradian** [comment]
By dangerous I meant that the consequences of bugs can be very bad, I wasn't stating the probability of that happening. But, if I have to, then I can add that it's not just a function of LVM code quality. There's also the probability of Salt having bugs and the probability of Salt user's code having bugs. Some of these numbers are coming from pillars and some pillar files are a maze of twisty little ifs which make it very hard to figure out what would happen if something is changed. Not your direct problem, but it would be nice to have guards which would prevent your users from shooting themselves in the foot.

And Linux LVM itself can be buggy if you attempt to create not so common things, like snapshots or thin LVs or change options from their defaults.

**Comment by cmcmarrow** [comment]
Thanks @piterpunk for your input. I think we should change this issue from bug to feature request. Then we can appropriately restructure lvm so @dkacar-oradian can have the functionality wanted and also have it be properly supported.

**Comment by dkacar-oradian** [comment]
I don't think this should be changed to feature request. I'm quite frustrated with bugs being reclassified as feature requests. ("Breaks existing functionality" was the criteria for RC bugs, IIRC).

But that's just frustrating and isn't the real issue. I need to know if something's going to be done about this before the release (which was scheduled for today, if I'm not mistaken) or not. If not, then I can't upgrade. Which isn't really all that good.

So, if not, I'm going to support my self by turning all this into a statefull script. This is in a macro anyway and I have exactly one place to change and then everything will work. And if I go that way, you don't have to spend the time on feature request.

But I need to know what's it gonna be.

**Comment by cmcmarrow** [comment]
@dkacar-oradian I feel like this issue can go either way because the docs say’s it only takes an `int` `. It could have just been dumb luck that it worked with `str` or that the function arg was miss documented. But like you said this bug breaks broken functionality that people rely on. So I have to agree with you.

I got this issue approved for the next point release! I hope you have a good day!

**Comment by piterpunk** [comment]
Nice.

I already wrote the patch and the behaviour will be:

- If the Logical Volume does not exist, create it passing the 'N%FREE', 'N%VG', etc, directly to linux_lvm.lvcreate
- If the Logical Volume exists and the existent % is different than the actual, it will do nothing.
- If the amount of extents is a natural number the logical volume will be created or resized to the specified, "force" option needs to be "true" to shrink a volume.

AFAIK that's also what the Puppet LVM module does. Ansible does a different approach, it calculates the amount of extents of %VG, %PVS and %FREE and resize accordingly, taking care that the "force" option was present when shrinking the Logical Volume.

This route is OK?

**Comment by steverweber** [comment]
Thanks for the quick work @piterpunk

**Comment by piterpunk** [comment]
> Thanks for the quick work @piterpunk

Maybe too quick, the "extents" parameter description got truncated, fixed now.
